pkg/storage/mvcc.go

```diff
@@ -4301,6 +4361,305 @@ func MVCCGarbageCollect(
+			LowerBound: gcKey.LatchSpan.Key,
+			UpperBound: gcKey.LatchSpan.EndKey,
+			KeyTypes:   IterKeyTypeRangesOnly,
+		})
+
+		iter.SeekGE(MVCCKey{Key: gcKey.LatchSpan.Key})
+
+		for ; ; iter.Next() {
+			if ok, err := iter.Valid(); err != nil {
+				return err
+			} else if !ok {
+				break
+			}
+
+			bounds := iter.RangeBounds()
+			unsafeRangeKeys := iter.RangeKeys()
+
+			// Check if preceeding range tombstone is adjacent to GC'd one. If we
+			// started iterating too early, just skip to next key. If boundaries
+			// match, then we capture start of the range and timestamps for later.
+			switch bounds.EndKey.Compare(gcKey.StartKey) {
+			case -1:
+				continue
+			case 0:
+				mergeTracker.update(bounds.Key, bounds.EndKey, unsafeRangeKeys)
+				continue
+			}
+
+			// Terminate loop once we've reached a range tombstone past the right
+			// GC range key boundary.
+			if cmp := bounds.Key.Compare(gcKey.EndKey); cmp >= 0 {
+				mergeTracker.update(bounds.Key, bounds.EndKey, unsafeRangeKeys)
+				break
+			}
+
+			// Check if we have a partial overlap between range tombstone and
+			// requested GCd range. This shouldn't happen in most cases, but we can
+			// have a range merge between GC run and cmd_gc execution or
+			// erroneous GC request and mvcc stats should be updated correctly.
+			// In those cases gcKey boundaries will not match underlying range
+			// tombstone boundaries and need to be adjusted.
+			trimLeft, trimRight := false, false
+			candidateStartKey := bounds.Key
+			if gcKey.StartKey.Compare(candidateStartKey) > 0 {
+				candidateStartKey = gcKey.StartKey
+				trimLeft = true
+			}
+			candidateEndKey := bounds.EndKey
+			if gcKey.EndKey.Compare(candidateEndKey) < 0 {
+				candidateEndKey = gcKey.EndKey
+				trimRight = true
+			}
+
+			removedRange := MVCCRangeKeyValue{
+				RangeKey: MVCCRangeKey{
+					StartKey: candidateStartKey,
+					EndKey:   candidateEndKey,
+				},
+			}
+			remaining := len(unsafeRangeKeys)
+			for i, rkv := range unsafeRangeKeys {
+				removedRange.RangeKey.Timestamp = rkv.RangeKey.Timestamp
+				removedRange.Value = rkv.Value
+				remove := rkv.RangeKey.Timestamp.LessEq(gcKey.Timestamp)
+				if remove {
+					if err := rw.ClearMVCCRangeKey(removedRange.RangeKey); err != nil {
+						return err
+					}
+					remaining--
+				}
+				if ms != nil {
+					ms.Add(updateStatsOnRangeTombstoneGC(removedRange, trimLeft, trimRight, i == 0, remove))
+				}
+			}
+
+			mergeTracker.update(candidateStartKey, candidateEndKey, unsafeRangeKeys[0:remaining])
+
+			// Verify that there are no remaining data under the deleted range using
+			// time bound iterator.
+			ptIter = NewMVCCIncrementalIterator(rw, MVCCIncrementalIterOptions{
+				KeyTypes:     IterKeyTypePointsOnly,
+				StartKey:     gcKey.StartKey,
+				EndKey:       gcKey.EndKey,
+				EndTime:      gcKey.Timestamp,
+				IntentPolicy: MVCCIncrementalIterIntentPolicyEmit,
+			})
+			ptIter.SeekGE(MVCCKey{Key: candidateStartKey})
```

Comment:
Looks like bounds were too broad, I should have used the bound of trimmed down range here instead. There's a room to optimize that to reduce number of seeks if range fragments are sequential, but I think it is not worth if we don't expect multiple fragments in general. Otherwise we can do that here and in range merging parts.